src/main/c/NativeModule.cpp

```diff
@@ -0,0 +1,828 @@
+    //  to other processes throughout the renaming operation and refer either
+    //  to the file referred to by new or old before the operation began."
+    //
+    // Since perl has a nasty habit of doing funny things to some characters
+    // in non-quoted strings, we take care to quote the filenames.
+    //
+    // It's enormously confusing to be writing C++ to write a makefile
+    // which contains embedded bash scripts which call perl.  That was not
+    // the original design, but issues of MacOS-vs-Linux portability left
+    // only a narrow path to a working solution.
+    FILE* f = fopen(hm_mak_.c_str(), "w");
+    if (!f) { perror("fopen"); return; }
+    std::string javaHome = config.java_home_;
+    std::string javaMD = config.java_md_;
+    fprintf(f, "MODULE    := hm_%s\n", key_.c_str());
+    fprintf(f, "MODULE_SO := $(MODULE)%s\n", config.ext_lib_.c_str());
+    fprintf(f, "PERL      := %s\n", config.perl_name_.c_str());
+    fprintf(f, "CXX       := %s\n", config.cxx_name_.c_str());
+    fprintf(f, "CXXFLAGS  := \\\n");
+    fprintf(f, "  %s -fPIC -march=native -fno-strict-aliasing -Wall \\\n", config.cxx_std_.c_str());
+    auto java_include = strip_suffix(config.java_home_, "/jre") + "/include";
+    fprintf(f, "  -I%s \\\n", java_include.c_str());
+    fprintf(f, "  -I%s/%s \\\n", java_include.c_str(), config.java_md_.c_str());
+    fprintf(f, "  -I%s \\\n", include_.c_str());
+    bool have_oflag = (strstr(options_.c_str(), "-O") != nullptr);
+    fprintf(f, "  %s%s \\\n", have_oflag ? "" : "-O3", options_.c_str());
+    fprintf(f, "  -DHAIL_MODULE=$(MODULE)\n");
+    fprintf(f, "LIBFLAGS := -fvisibility=default %s\n", 
+      config.is_darwin_ ? "-dynamiclib -Wl,-undefined,dynamic_lookup"
+                         : "-rdynamic -shared");
+    fprintf(f, "\n");
+    // top target is the .so
+    fprintf(f, "$(MODULE_SO): $(MODULE).o\n");
+    fprintf(f, "\t$(PERL) -e 'rename \"$(MODULE).new\", \"$@\"'\n");
+    fprintf(f, "\n");
+    // build .o from .cpp
+    fprintf(f, "$(MODULE).o: $(MODULE).cpp\n");
+    fprintf(f, "\t$(CXX) $(CXXFLAGS) -o $@ -c $< 2> $(MODULE).err && \\\n");
+    fprintf(f, "\t  $(CXX) $(CXXFLAGS) $(LIBFLAGS) -o $(MODULE).tmp $(MODULE).o 2>> $(MODULE).err ; \\\n");
+    fprintf(f, "\tstatus=$$? ; \\\n");
+    fprintf(f, "\tif [ $$status -ne 0 ] || [ -z $(MODULE).tmp ]; then \\\n");
+    fprintf(f, "\t  /bin/rm -f $(MODULE).new ; \\\n");
+    fprintf(f, "\t  echo FAIL ; exit 1 ; \\\n");
+    fprintf(f, "\tfi\n");
+    fprintf(f, "\t-/bin/rm -f $(MODULE).err\n");
+    fprintf(f, "\t$(PERL) -e 'rename \"$(MODULE).tmp\", \"$(MODULE).new\"'\n");
+    fprintf(f, "\n");
+    fclose(f);
+  }
+
+public:
+  bool try_to_start_build() {
+    // Try to create the .new file
+    std::lock_guard<NativeModule> mylock(*parent_);
+    int fd = ::open(hm_new_.c_str(), O_WRONLY|O_CREAT|O_EXCL, 0666);
+    if (fd < 0) {
+      // We lost the race to start the build
```

Comment:
There's no race if have your own /tmp directory and you're locking in process.

Reply:
You could redesign it to eliminate the reliance on O_EXCL, but it would require more far-reaching changes - for this particular case, the thread would need to hold the from before where it tests for the existence of hm_new_, through to at least to after it creates the file.

Reply:
See my comment above.  The master where the compilation happens should all be single threaded.